### PR TITLE
Additions to MeshBuilder3D +

### PR DESCRIPTION
**Pull request by dhakehurst**
Add a 'shape3D' to the DSL.
add funs to MeshBuilder3D for cuboid, cube, sphere, rect, triangle
begin support for adding text to 3D shapes.
fix typo shiness --> shininess

### CODE REVIEW

**Comment by soywiz** [comment]
Thanks. Very nice contribution. I'll check asap

**Comment by soywiz** [comment]
Thanks for your patience, I had to do lots of chores after the holidays. Checking already

File: korge/src/commonMain/kotlin/com/soywiz/korge3d/BatchBuilder3D.kt [hunk 1]
```diff
@@ -0,0 +1,497 @@
...
+ * Allows to draw quads and sprites buffering the geometry to limit the draw batches executed calling [AG] (Accelerated Graphics).
+ * This class handles a vertex structure of: x, y, u, v, colorMul, colorAdd. Allowing to draw texturized and tinted primitives.
+ *
+ * You should call: [drawQuad], [drawQuadFast], [drawNinePatch], [drawVertices] for buffering the geometries.
+ *
+ * For performance the actual drawing/rendering doesn't happen until the [flush] method is called (normally that happens automatically).
+ * Then the engine will call [flush] when required, automatically once the buffer is filled, at the end of the frame, or when [RenderContext.flush] is executed
+ * by other renderers.
+ */
+class BatchBuilder3D(
```
Comment (soywiz) [inline korge/src/commonMain/kotlin/com/soywiz/korge3d/BatchBuilder3D.kt:28]: Why is this class required? Seems a copy of BatchBuilder2D. Since I cannot get the diff, which are the differences, and why is this needed?

File: korge/src/commonMain/kotlin/com/soywiz/korge3d/Library3D.kt [hunk 2]
```diff
@@ -204,7 +204,7 @@ fun Library3D.MaterialDef.instantiate(): Material3D {
 		ambient = effect?.ambient?.instantiate() ?: Material3D.LightColor(Colors.BLACK),
 		diffuse = effect?.diffuse?.instantiate() ?: Material3D.LightColor(Colors.BLACK),
 		specular = effect?.specular?.instantiate() ?: Material3D.LightColor(Colors.BLACK),
-		shiness = effect?.shiness ?: 0.5f,
+        shininess = effect?.shiness ?: 0.5f,
```
Comment (soywiz) [inline korge/src/commonMain/kotlin/com/soywiz/korge3d/Library3D.kt:207]: I guess you should rename `effect?.shiness` too

File: korge/src/commonMain/kotlin/com/soywiz/korge3d/BitmapFontExt3D.kt [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+import com.soywiz.korma.geom.Matrix
+import com.soywiz.korma.geom.Vector3D
+import com.soywiz.korma.geom.invert
+
+fun BitmapFont.drawText3D(
+    ctx: RenderContext3D,
+    v1: Vector3D,
+    v2: Vector3D,
+    v3: Vector3D,
+    v4: Vector3D,
```
Comment (soywiz) [inline korge/src/commonMain/kotlin/com/soywiz/korge3d/BitmapFontExt3D.kt:21]: What is each vector for? Either a small documentation line explaining them, or if there are other names than vN to understand these parameters would be nice

Comment (soywiz) [reply korge/src/commonMain/kotlin/com/soywiz/korge3d/BitmapFontExt3D.kt:21]: As I checked later, this might be the face containing the text? Maybe we can create a class for the Face? Also, could we determine a quad in the 3D space with only three points and compute the other one? Maybe a Face class could allow to construct that instance several ways and provide it here.

File: korge/src/commonMain/kotlin/com/soywiz/korge3d/MeshBuilder3D.kt [hunk 4]
```diff
@@ -1,42 +1,203 @@
...
+import com.soywiz.korim.color.RGBA
+import com.soywiz.korim.font.BitmapFont
+import com.soywiz.korio.util.buildList
+import com.soywiz.korma.geom.Vector3D
+import kotlin.math.PI
+import kotlin.math.cos
+import kotlin.math.sin
 
 @Korge3DExperimental
 class MeshBuilder3D {
```
Comment (soywiz) [inline korge/src/commonMain/kotlin/com/soywiz/korge3d/MeshBuilder3D.kt:18]: Nice additions!

File: korge/src/commonMain/kotlin/com/soywiz/korge3d/Shapes3D.kt [hunk 5]
```diff
@@ -1,15 +1,47 @@
...
 
 @Korge3DExperimental
 @Deprecated("Kotlin/Native boxes inline+Number")
-inline fun Container3D.box(width: Number, height: Number, depth: Number, callback: Cube3D.() -> Unit = {}): Cube3D
-    = box(width.toDouble(), height.toDouble(), depth.toDouble(), callback)
+inline fun Container3D.cube(width: Number, height: Number, depth: Number, callback: Cube3D.() -> Unit = {}): Cube3D = cube(width.toDouble(), height.toDouble(), depth.toDouble(), callback)
 
 @Korge3DExperimental
-inline fun Container3D.box(
+inline fun Container3D.cube(
```
Comment (soywiz) [inline korge/src/commonMain/kotlin/com/soywiz/korge3d/Shapes3D.kt:44]: We have several terms: `box`, `cube` and `cuboid`. It seems that other programs including blender uses cube so the change makes sense. But in the MeshBuilder3D looks like there is a cuboid method. Is that a normal cube/box, or is something else?

**Comment by dhakehurst** [comment]
no worries

File: korge/src/commonMain/kotlin/com/soywiz/korge3d/BatchBuilder3D.kt (hunk 1 above)
Comment (dhakehurst) [reply korge/src/commonMain/kotlin/com/soywiz/korge3d/BatchBuilder3D.kt:28]: I think it is no longer needed.
I was using it as part of the 3D Text impl.
But have had to change the approach.
I did say that the 3DText is still WIP. (maybe it should be on a separate PR)

File: korge/src/commonMain/kotlin/com/soywiz/korge3d/Library3D.kt (hunk 2 above)
Comment (dhakehurst) [reply korge/src/commonMain/kotlin/com/soywiz/korge3d/Library3D.kt:207]: yes :-)

File: korge/src/commonMain/kotlin/com/soywiz/korge3d/Shapes3D.kt (hunk 5 above)
Comment (dhakehurst) [reply korge/src/commonMain/kotlin/com/soywiz/korge3d/Shapes3D.kt:44]: a cube has width == height == depth.
if they are different it is called a cuboid.
(akin to square and rectangle in 2D)

File: korge/src/commonMain/kotlin/com/soywiz/korge3d/BitmapFontExt3D.kt (hunk 3 above)
Comment (dhakehurst) [reply korge/src/commonMain/kotlin/com/soywiz/korge3d/BitmapFontExt3D.kt:21]: good idea. 
see above comment re 3D Text being in progress

File: korge/src/commonMain/kotlin/com/soywiz/korge3d/Shapes3D.kt (hunk 5 above)
Comment (soywiz) [reply korge/src/commonMain/kotlin/com/soywiz/korge3d/Shapes3D.kt:44]: Yaah, that's what I thought. I guess I called `box` (though probably `cuboid` more correct) because of that. But it seems that programs use `cube` for simplicity. Or that initially has all its components to 1,1,1, and you use transforms to deform it converting it into a cuboid.

File: korge/src/commonMain/kotlin/com/soywiz/korge3d/BatchBuilder3D.kt (hunk 1 above)
Comment (soywiz) [reply korge/src/commonMain/kotlin/com/soywiz/korge3d/BatchBuilder3D.kt:28]: I see. If this is not required anymore, maybe this can be removed.

Or if you need something very similar to BatchBuilder2D, we can rename BatchBuilder2D to just BatchBuilder, then make that open or abstract so we can reuse the code. Alternatively if we just need some extension methods, maybe the BatchBuilder could be closed, which usually gives better performance on native.

File: korge/src/commonMain/kotlin/com/soywiz/korge3d/BitmapFontExt3D.kt (hunk 3 above)
Comment (soywiz) [reply korge/src/commonMain/kotlin/com/soywiz/korge3d/BitmapFontExt3D.kt:21]: As you prefer. If you think this PR without the text additions makes sense already, we can merge it, so you can add the text one in a separate PR.

Comment (dhakehurst) [reply korge/src/commonMain/kotlin/com/soywiz/korge3d/BitmapFontExt3D.kt:21]: the 3D Text is proving to be more work than I initially though, so it is probably best to keep that to a diff PR.
If it is easy for you to pick out the shiness typo fixes and MeshBuilder stuff, then do that from this PR.
Else close it and I will try and create a diff branch without the 3D Text bits, and give you a new PR from that.

Comment (soywiz) [reply korge/src/commonMain/kotlin/com/soywiz/korge3d/BitmapFontExt3D.kt:21]: It's Ok. This repo doesn't need to be super stable right now. I can merge as it is right now, so we don't have to rebase stuff

**Comment by soywiz** [comment]
Thanks!

File: korge/src/commonMain/kotlin/com/soywiz/korge3d/BitmapFontExt3D.kt (hunk 3 above)
Comment (dhakehurst) [reply korge/src/commonMain/kotlin/com/soywiz/korge3d/BitmapFontExt3D.kt:21]: ok. great. I am will continue the 3D Text, and give a separate PR when I have something (more) working.
